gov/politeia/proposals.go

```diff
@@ -91,373 +90,607 @@ func NewProposalsDB(politeiaURL, dbPath string) (*ProposalDB, error) {
-	return &pInfo, nil
+	// Iterate through every record and feed data used by dcrdata
+	proposals := make([]*pitypes.ProposalRecord, 0, len(recordDetails))
+	for _, record := range recordDetails {
+		// Record data
+		proposal := &pitypes.ProposalRecord{
+			State:     record.State,
+			Status:    record.Status,
+			Version:   record.Version,
+			Timestamp: uint64(record.Timestamp),
+			Username:  record.Username,
+			Token:     record.CensorshipRecord.Token,
+		}
+
+		// Proposal metadata
+		pm, err := proposalMetadataDecode(record.Files)
+		if err != nil {
+			return nil, fmt.Errorf("proposalMetadataDecode err: %w", err)
+		}
+		proposal.Name = pm.Name
+
+		// User metadata
+		um, err := userMetadataDecode(record.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("userMetadataDecode err: %w", err)
+		}
+		proposal.UserID = um.UserID
+
+		// Comments count
+		commentsCount, ok := commentsCounts[proposal.Token]
+		if !ok {
+			log.Errorf("Comments count for proposal %v not returned by API",
+				proposal.Token)
+			continue
+		}
+		proposal.CommentsCount = int32(commentsCount)
+
+		// Vote summary data
+		summary, ok := voteSummaries[proposal.Token]
+		if !ok {
+			log.Errorf("Vote summary for proposal %v not returned by API",
+				proposal.Token)
+			continue
+		}
+		proposal.VoteStatus = summary.Status
+		proposal.VoteResults = summary.Results
+		proposal.EligibleTickets = summary.EligibleTickets
+		proposal.StartBlockHeight = summary.StartBlockHeight
+		proposal.EndBlockHeight = summary.EndBlockHeight
+		proposal.QuorumPercentage = summary.QuorumPercentage
+		proposal.PassPercentage = summary.PassPercentage
+
+		var totalVotes uint64
+		for _, v := range summary.Results {
+			totalVotes += v.Votes
+		}
+		proposal.TotalVotes = totalVotes
+
+		// Status change metadata
+		statusTimestamps, changeMsg, err :=
+			statusChangeMetadataDecode(record.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("statusChangeMetadataDecode err: %w", err)
+		}
+		proposal.PublishedAt = uint64(statusTimestamps.published)
+		proposal.CensoredAt = uint64(statusTimestamps.censored)
+		proposal.AbandonedAt = uint64(statusTimestamps.abandoned)
+		proposal.StatusChangeMsg = changeMsg
+
+		// Append proposal after inserting the relevant data
+		proposals = append(proposals, proposal)
+	}
+
+	return proposals, nil
 }
 
-// LastProposalsSync returns the last time a sync to update the proposals was run
-// but not necessarily the last time updates were synced in proposals.db.
-func (db *ProposalDB) LastProposalsSync() int64 {
-	return atomic.LoadInt64(&db.lastSync)
+// fetchVettedTokens fetches all vetted tokens ordered by the timestamp of
+// their last status change.
+func (db *ProposalsDB) fetchVettedTokensInventory() ([]string, error) {
+	page := 1
+	var vettedTokens []string
+	for {
+		inventoryReq := recordsv1.InventoryOrdered{
+			State: recordsv1.RecordStateVetted,
+			Page:  uint32(page),
+		}
+		reply, err := db.client.RecordInventoryOrdered(inventoryReq)
```

Comment:
What would be returned for a non-existent/empty page?  This is an important question since the condition below does not guarantee it is the last page since the last page can also have `len(reply.Tokens) == int(recordsv1.InventoryPageSize)`.

Reply:
Yep important question. The request goes through fine for an empty page, no errors returned. The tokens slice of the reply will return empty though, so in the case of the last page being full, the next request would fetch an empty token slice and we'd know that it was indeed the last page. Should I doc this?

Reply:
Can't hurt to doc intent and assumptions whenever it's not obvious.